docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+
+1. In our code, there're multiple places where we use `tokio::watch::{Sender, Receiver}` with type `()`. It's hard to track down whether we're just waiting for the drop, for a signal or both from the sender. To make it clear, a couple of wrappers are needed:
+    - `SignalSender` / `SignalWatcher`: wait only for proper signals and `panic!` when it receives `RecvError`.
+    - `DropSender` / `DropWatcher`: wait only for `RecvError` and don't have any function in the interface to signal.
+    - Non-generic `Sender` / `Receiver` wrappers for both drops and signals.
+2. For the cancellation proposal, we'll need an object that both can wait for a task to finish and send a cancellation request. We'll call it `CancelationSender` / `CancellationReceiver`. It could be coded using two tokio channels.
+3. For the cancellation proposal, we'll need a triple `StateKeeper` / `StateSubscriber` / `StateHider`. This triple atomically changes the state and sends it to subscribers like `tokio::watch`. But when there's at least one `StateHider`, we put the new state to the buffer that is being read only by subscribers, and `StateKeeper` will read the older value. After all `StateHider`s are gone, we atomically (with respect to the new possible `StateHider`s) change the state.
+
+    ```rust
+    impl<S> StateKeeper<S> {
+        pub fn get_state(&self) -> S;
+        pub fn set_state(&self, state: S);
+        pub fn subscribe(&self) -> StateSubscriber;
+    }
+
+    impl<S> StateSubscriber<S> {
+        pub fn get_state(&self) -> S;
+        pub fn state_hider(&self) -> StateHider;
+    }
+
+    impl<S> Clone for StateSubscriber<S>;
+
+    impl<S> Drop for StateHider<S>;
+    ```
+
+ The purpose of this non-trivial primitive is to hide the state changes from the closures running in the tenant and timeline until closures finish the work. It is explained in more detail later.
+
+## States and transitions in `TenantState`
+
+Currently, we have 3 tenant states:
+
+- `Active { background_jobs_running: bool }` - fully operational, _its background jobs might be running or not_.
+- `Paused` - is recognised by the pageserver but not yet ready to operate, e.g. not present locally and being downloaded or being read into memory from the file system.
+- `Broken` - is recognised by the pageserver but no longer used for any operations, as it failed to get activated.
+
+Problems:
+
+- It needs to be clarified how to know that tenant is being dropped.
+- `Active` state has an option that's not needed. If we do not want to run background tasks, we may wish to have an alternative to the `Active` state without running background jobs.
+- Actually, we can change the state from `Paused` to `Active` _and vice versa_! Engineers must remember that when writing the code, the only reason to get to know that is to read the code. Only some of the people in the team know about that.
+
+Proposed states:
+
+1. `Infant` - just created, no activity yet.
```

Comment:
Looks redundant to me. May be needed only if we want to have some dummy initial state for Watcher so subscribers will be able observe Infant -> Loading change.

Reply:
I don't know here. But probably, one more state won't hurt, as it shows useful information.